nexus/src/sagas.rs

```diff
@@ -1289,6 +1289,78 @@ async fn sdc_regions_ensure(
 
     let block_size = datasets_and_regions[0].1.block_size;
 
+    // If requested, back disk by image
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let read_only_parent: Option<
+        Box<sled_agent_client::types::VolumeConstructionRequest>,
+    > = if let Some(image_id) = params.create_params.image_id {
+        let log = osagactx.log();
+        warn!(log, "grabbing image {}", image_id);
+
+        // TODO: support project images too
+        let (.., global_image) = LookupPath::new(&opctx, &osagactx.datastore())
+            .global_image_id(image_id)
+            .fetch()
+            .await
+            .map_err(ActionError::action_failed)?;
+
+        debug!(log, "retrieved global image {}", global_image.id());
+
+        // If the block size of the requested disk doesn't match the image,
+        // return an error
+        if global_image.block_size.to_bytes() as u64
+            != params.create_params.block_size().to_bytes()
```

Comment:
(brainstorming) Would it be possible to structure the API to avoid this conflict?

- The "DiskCreate" parameters supply block size
- The "ImageCreate" parameters supply block size
- The two must match
- ... but if someone is calling "DiskCreate", and pointing to an image, the block size is basically implied

The current format of DiskCreate is fairly flat - a user can supply an optional "snapshot ID", or an optional "image ID", or neither, (or both?) and must *always* supply the block size.

But if we updated it to use an enum like:

```rust
enum DiskSource {
  Blank(BlockSize),
  Snapshot(Uuid),
  Image(Uuid),
}

pub struct DiskCreate {
  pub identity: IdentityMetadataCreateParams,
  pub source: DiskSource,
  pub size: ByteCount,
}
```

That might avoid this error condition by construction.

Reply:
I like this - done in 0477683b